### ISSUE TITLE
Kivy blurs pixels in Windows when UI scaling is enabled

### ISSUE THREAD

**Issue by BenMcLean** [kivy/kivy#3705]
In Windows 10, my resolution is 1920x1080 and UI scaling is set to 175%. With these settings, (and possibly others) the pixels in kivy apps appear blurred around the edges. Not sure if this is true across all Windows versions or just Windows 10.

Talking about this in IRC led me to discover that using kivy.metrics.dp(1) returns 1.0 and should ((unless I am mistaken) return 1.75 in this case.

Kivy needs to be made sensitive to where that information is exposed in the Windows API to fix this, so that this type of scaling can be applied with the nearest neighbor algorithm to keep the pixels sharp.

I can supply code that reproduces this issue if it will help.

As a temporary workaround, I have checked the "Disable display scaling on high DPI settings" option in the Windows Compatibility settings on my python.exe for all users. That results in sharp unscaled pixels. It would still be nice to get this fixed so we can have sharp scaled pixels.

**Comment by BillBridge** [comment]
I am having the same problems. See http://stackoverflow.com/questions/38049695/kivy-gui-does-not-adapt-to-customize-display-on-windows?noredirect=1#comment63553443_38049695

What is the status for Windows as it seems to be resolved for OsX?

**Comment by verata-veritatis** [comment]
Still no updates huh? Having issues with this on Win10 currently.

**Comment by kuguma** [comment]
I also encountered the same problem on Win10. From the metrics.py dpi acquisition code, it seems that the code for windows is not implemented yet.

In conclusion, I was able to solve this problem via ctypes.
My code gets the window handle from tkinter and gets the DPI from there.
It would be better to get the handle from SDL. But I don't know how.
If someone tells me, I can send a pull request.

**Comment by dangpzanco** [comment]
This is still a problem after +4 years....

**Comment by CoziestYew804** [comment]
@kuguma can you please explain how you did that ?

**Comment by probablyfine** [comment]
I've got this blurry pixel issue as well on Windows 10 with UI scaling at 150%. Another related problem is that the Window core module reports the wrong size/position.

@kuguma is correct, on Windows the metrics.py code always returns 1.0 in its `density` method. Based on their suggestion to solve this using ctypes (plus some googling), I hacked together a fix. I don't know how "proper" this is; there could be issues I'm not aware of and I don't know if it will work in every environment. Hopefully someone who understands this better can weigh in and let me know if this is a reasonable approach.

In any case, it works great on my system as far as I can tell: text is perfectly crisp, and the Window module reports the correct size and position. I checked that it works across all the UI scaling options I have available, but obviously the app needs to be restarted after a setting change. I do not know how this will work if you're using multiple displays.

To apply the hack, edit your [metrics.py file and delete line 218](https://github.com/kivy/kivy/blob/3e37301b291cfbd8e4cde5f1d0e7ce739700d8da/kivy/metrics.py#L218). Replace it with these lines:
`        from win32api import GetSystemMetrics
        from ctypes import windll
        w0 = GetSystemMetrics(0)
        h0 = GetSystemMetrics(1)
        windll.user32.SetProcessDPIAware()
        w1 = GetSystemMetrics(0)
        h1 = GetSystemMetrics(1)
        sf_w = w1/w0
        sf_h = h1/h0
        return min( sf_w, sf_h)`

**Comment by matham** [comment]
Can you try https://github.com/kivy/kivy/pull/7293 ? That should allow you to set a custom density, or make the window report the correct density from sdl2.

**Comment by probablyfine** [comment]
@matham thanks very much for pointing me towards this, I hadn't seen it. Will give it a try as soon as I get a chance.

**Comment by probablyfine** [comment]
@matham I'm having trouble getting the new features in the PR you linked to work properly. I copied your demo code, and I am able to see some effect of dynamic scaling using this line: `Clock.schedule_once(partial(set_density, 1.5), 15)`

I think that should set the density to 1.5 instead of the default 1.0, but even though there is an obvious visual change when the `Clock` timer runs out, the result doesn't look "right"-- text is still as blurry as in the master branch. The values in `Window.size` also don't seem to change along with the visual change. Using the hack I posted, I get crisp text with density set to 1.5 and `Window.size` changes to the correct values.

Could you explain how to get the window to report the correct density from sdl2? (This is the feature I'm mainly after.)

**Comment by matham** [comment]
Ok, I see the issue now. By default Windows thinks Kivy is not DPI aware so Windows scales it up which makes it look blurry. Your code tell Windows that the window is dpi aware so Windows doesn't scale Kivy, rather you set a higher density and that does the scaling. So it looks nice.

The PR I linked changes only the density, but it does not tell Windows to not scale the window, hence it looks blurry. We'd have to also tell the Window that we're DPI aware and to not scale us. Normally, we'd go through SDL2 to do this, as that's what we use to control the window. However, SDL2 doesn't seem to support DPI scaling for Windows yet AFAICT (https://bugzilla.libsdl.org/show_bug.cgi?id=3281).

So I think your workaround is the best solution until SDL2 gains DPI awareness on Windows and allows us to read it or set it.

**Comment by probablyfine** [comment]
Thanks a lot for those clarifying details, I think I understand better about what's going on here. I'm still unsure how multiple monitors with different DPIs might interact with the ctypes fix-- probably just will be too big or too small on one monitor? I may get a chance to try it out sometime soon and if so will post back here about how it goes.

**Comment by matham** [comment]
>I'm still unsure how multiple monitors with different DPIs might interact with the ctypes fix-- probably just will be too big or too small on one monitor?

Yes. If the monitors have different scaling factors, I think that'll be the case. Although there are two dpi-aware Windows modes, one uses the DPI of the first monitor created on and Windows will always report the same density once created (that's what you're setting above), and the other updates per-monitor. However, unlike if we get the event from SDL2 and we could just react to that event, you'd need to check manually on every clock frame to see if the density has changed because the monitor has changed, and update the density accordingly. And you'd have to use the linked PR, because with master density cannot be changed once it's set.

**Comment by probablyfine** [comment]
I did a bit more reading on this, it looks to me like the "best"/most-modern way to set DPI awareness per-monitor (rather than just the first monitor) is with: `windll.user32.SetProcessDpiAwarenessContext(-4)`. The `-4` apparently signifies `DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2`, which is the newest iteration of this type of context-awareness. The calls to `win32api.GetSystemMetrics` can be replaced with `windll.user32.GetSystemMetricsForDpi` for the per-monitor version. I don't have a second monitor handy at the moment but these appear to work at least on a single monitor. The downside is these functions are only supported on Windows 10 Creators Update (1703) and newer.

The info [here](https://docs.microsoft.com/en-us/windows/win32/hidpi/high-dpi-desktop-application-development-on-windows) and [here](https://docs.microsoft.com/en-us/windows/win32/hidpi/setting-the-default-dpi-awareness-for-a-process) says there is also a version of this procedure for Windows 8.1 and newer, as well as a version for Windows Vista and newer (this is the one I had posted above, so not per-monitor aware). I think it would be pretty simple to have kivy check the OS version and use the correct one, falling back on density=1.0 for older versions?

In terms of reacting to the DPI change (using your new features in your PR), it also [looks like](https://docs.microsoft.com/en-us/windows/win32/hidpi/wm-dpichanged) Windows will send the window a WM_DPICHANGED message if it gets moved between monitors with different DPIs (if the per-monitor awareness mode is set). I wonder if kivy can react to this event like other types of events, and trigger the update?

[This person](https://stackoverflow.com/questions/55210980/not-receiving-wm-dpichanged-in-sdl-application) is doing some of this, and happens to be using SDL2 as well. They say that SDL2 for some reason consumed the WM_DPICHANGED message and didn't pass it along, though I'm not familiar enough with kivy's event processing methods to know if that would be relevant here, too. That person did find a workaround, though (but again not sure if theirs would work here).

**Comment by tshirtman** [comment]
kivy's management of windows event is there,
https://github.com/kivy/kivy/blob/master/kivy/core/window/_window_sdl2.pyx#L557
experimenting with it requires compilation, if only to uncomment https://github.com/kivy/kivy/blob/master/kivy/core/window/_window_sdl2.pyx#L659 and see what's printed, but that should allow testing that hypothesis quickly.

It would probably be fine to have some os-specific code to handle setting the proper options at window creation.

**Comment by probablyfine** [comment]
Thanks for that tip @tshirtman ! I'll give it a try when I get a chance. I'll also try to grab a second monitor as well and see what I can see.

**Comment by probablyfine** [comment]
Unfortunately no luck here: when I uncomment line 659 in the file you linked (as well as line 675), and set a test app to be DPI aware, I don't get any events printed to console when I change UI scaling settings. I assume the same would be true for dragging the window to a second monitor but I haven't tested that.

The change to _window_sdl2.pyx does appear to work at least: I get "receive unknown sdl window event 512" printed to console when I manually resize the window (not sure why this is an unknown event). So that makes me think I've at least set things up correctly for this test.

In the stackoverflow thread I posted before, the OP left [this](https://stackoverflow.com/questions/55210980/not-receiving-wm-dpichanged-in-sdl-application#comment97189702_55210980) comment suggesting they solved this by "subclassing the SDL window". I followed the link they left on that approach but I don't understand it well and I'm not sure if something similar could be done here.

Thinking through @matham's previous suggestion to manually check for a DPI change on each clock cycle: I guess the problem with that is an unacceptable hit to performance, or just generally being an ugly solution. But for something like this, I wonder if a check every 1 or 2 seconds would acceptable in terms of both responsive behavior and performance. Might give this a try...
